Add blog post for kube2pulumi (#3914)
pulumi/docs: content/blog/introducing-kube2pulumi/index.md

---
title: "Introducing kube2pulumi: No More YAML"

> date: 2020-08-12

Reviewer: ```suggestion
date: 2020-08-13
```

of manipulating YAML directly, you can take advantage of the rich ecosystem of programming language tools to
supercharge your productivity.

<!--more-->

Using programming languages to manage Kubernetes resources has many benefits: strong typing, linting, auto-completion,
refactoring, testing, and the ability to use standard language features like variables, loops, conditions, and
abstraction. Pulumi also supports [Helm], [Kustomize], and [YAML], but these tools don't take full advantage of the
programming language. With `kube2pulumi`, you can realize these benefits without manual conversion work upfront.

Check out an example of `kube2pulumi` in action, and then we'll take a closer look at the generated code.

## Getting Started with kube2pulumi

Let's start with a simple example: converting a Namespace. Here's the YAML manifest:

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: foo
```

`kube2pulumi` supports all of Pulumi's SDKs, so you can choose whatever language you like!

{{< chooser language "typescript,python,csharp,go" >}}

{{% choosable language typescript %}}
Here, we choose TypeScript and select our YAML manifest.

> ```

Reviewer 2: ```suggestion
```sh
```

$ kube2pulumi typescript -f namespace.yaml
Conversion successful! Generated File: namespace.ts
```

> ```ts

Reviewer 2: ```suggestion
```typescript
```

import * as pulumi from "@pulumi/pulumi";
import * as kubernetes from "@pulumi/kubernetes";

const fooNamespace = new kubernetes.core.v1.Namespace("fooNamespace", {
    apiVersion: "v1",
    kind: "Namespace",
    metadata: {
        name: "foo",
    },
});
```

{{% /choosable %}}

{{% choosable language python %}}
Here, we choose Python and select our YAML manifest.

> ```

Reviewer 2: ```suggestion
```sh
```

$ kube2pulumi python -f namespace.yaml
Conversion successful! Generated File: `__main__.py`
```

```python
import pulumi
import pulumi_kubernetes as kubernetes

foo_namespace = kubernetes.core.v1.Namespace("fooNamespace",
    api_version="v1",
    kind="Namespace",
    metadata={
        "name": "foo",
    })
```

{{% /choosable %}}

{{% choosable language csharp %}}
Here, we choose C# and select our YAML manifest.

> ```

Reviewer 2: ```suggestion
```sh
```

$ kube2pulumi csharp -f namespace.yaml
Conversion successful! Generated File: Program.cs
```

```csharp
using Pulumi;
using Kubernetes = Pulumi.Kubernetes;

class MyStack : Stack
{
    public MyStack()
    {
        var fooNamespace = new Kubernetes.Core.V1.Namespace("fooNamespace", new Kubernetes.Types.Inputs.Core.V1.NamespaceArgs
        {
            ApiVersion = "v1",
            Kind = "Namespace",
            Metadata = new Kubernetes.Types.Inputs.Meta.V1.ObjectMetaArgs
            {
                Name = "foo",
            },
        });
    }

}
```

{{% /choosable %}}

{{% choosable language go %}}
Here, we choose Go and select our YAML manifest.

> ```

Reviewer 2: ```suggestion
```sh
```

        metav1 "github.com/pulumi/pulumi-kubernetes/sdk/v2/go/kubernetes/meta/v1"
        "github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

func main() {
        pulumi.Run(func(ctx *pulumi.Context) error {
                _, err := corev1.NewNamespace(ctx, "fooNamespace", &corev1.NamespaceArgs{
                        ApiVersion: pulumi.String("v1"),
                        Kind:       pulumi.String("Namespace"),
                        Metadata: &metav1.ObjectMetaArgs{
                                Name: pulumi.String("foo"),
                        },
                })
                if err != nil {
                        return err
                }
                return nil
        })
}
```

{{% /choosable %}}

{{< /chooser >}}

## Multi-resource Example

That's pretty cool but doesn't show off the full power of `kube2pulumi`. Let's try a more realistic conversion
scenario. Our newly announced [Pulumi Kubernetes Operator] requires a few different resources to install: a Service
Account, Role, RoleBinding, and Deployment, totaling about 130 lines of YAML. Let's try converting that directory
to some real code!

> ```
> $ ls operator
> operator.yaml  role.yaml  role_binding.yaml  service_account.yaml
> 
> $ kube2pulumi typescript -d operator
> Conversion successful! Generated File: operator/main.ts
> ```

Reviewer 2: ````suggestion
```sh
$ ls operator
operator.yaml  role.yaml  role_binding.yaml  service_account.yaml
```

```sh
$ kube2pulumi typescript -d operator
Conversion successful! Generated File: operator/main.ts
```
````

    kind: "RoleBinding",
    apiVersion: "rbac.authorization.k8s.io/v1",
    metadata: {
        name: "pulumi-kubernetes-operator",
    },
    subjects: [{
        kind: "ServiceAccount",
        name: "pulumi-kubernetes-operator",
    }],
    roleRef: {
        kind: "Role",
        name: "pulumi-kubernetes-operator",
        apiGroup: "rbac.authorization.k8s.io",
    },
});
const pulumi_kubernetes_operatorServiceAccount = new kubernetes.core.v1.ServiceAccount("pulumi_kubernetes_operatorServiceAccount", {
    apiVersion: "v1",
    kind: "ServiceAccount",
    metadata: {
        name: "pulumi-kubernetes-operator",
    },
});
```

## Superpowers

Now that we've got some Kubernetes resources defined in real code, we can start working on them like any other software
project. First, we define a variable for the operator name and then swap out hardcoded string references to the other
resource names. With these changes, it doesn't require any thought to keep these resources in sync! You can later decide
to change the name in one place, and every other reference will update automatically as well.

> ```ts

Reviewer 2: ```suggestion
```typescript
```

---
title: "Introducing kube2pulumi: No More YAML"
date: 2020-08-13
meta_desc: Automatically convert Kubernetes YAML manifests to modern code in TypeScript, Python, C#, and Go.
meta_image: yaml2code.png
authors:
    - levi-blackstone
    - sashu-shankar
tags:
    - Kubernetes
---

Kubernetes users often joke about being "YAML engineers," and the pile of YAML seems to get deeper every day. Today,
we're pleased to announce [kube2pulumi], a tool to automatically convert Kubernetes manifests into modern code! Instead
of manipulating YAML directly, you can take advantage of the rich ecosystem of programming language tools to
supercharge your productivity.

<!--more-->

Using programming languages to manage Kubernetes resources has many benefits: strong typing, linting, auto-completion,
refactoring, testing, and the ability to use standard language features like variables, loops, conditions, and
abstraction. Pulumi also supports [Helm], [Kustomize], and [YAML], but these tools don't take full advantage of the
programming language. With `kube2pulumi`, you can realize these benefits without manual conversion work upfront.

> Check out an example of `kube2pulumi` in action, and then we'll take a closer look at the generated code.

Reviewer 3: Perhaps add a link to and image of pulumi.com/kube2pulumi here?  And then use the post to drill into details?  Now that we have the landing page, it's a great visual and engagement tool to recommend here.

FWIW - we may also want to add a link to that pulumi.com/kube2pulumi page back to this blog post once it's published - "want to learn more about kube2pulumi, check out the blog post for more details".

## Getting Started with kube2pulumi

Let's start with a simple example: converting a Namespace. Here's the YAML manifest:

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: foo
```

> `kube2pulumi` supports all of Pulumi's SDKs, so you can choose whatever language you like!

Reviewer 3: Not sure "Pulumi's SDKs " is clear here - suggestion below - but may need more rewording.

```suggestion
`kube2pulumi` supports all of Pulumi's supported languages, so you can choose whatever language you like!
```

        "github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

func main() {
        pulumi.Run(func(ctx *pulumi.Context) error {
                _, err := corev1.NewNamespace(ctx, "fooNamespace", &corev1.NamespaceArgs{
                        ApiVersion: pulumi.String("v1"),
                        Kind:       pulumi.String("Namespace"),
                        Metadata: &metav1.ObjectMetaArgs{
                                Name: pulumi.String("foo"),
                        },
                })
                if err != nil {
                        return err
                }
                return nil
        })
}
```

{{% /choosable %}}

{{< /chooser >}}

## Multi-resource Example

That's pretty cool but doesn't show off the full power of `kube2pulumi`. Let's try a more realistic conversion
scenario. Our newly announced [Pulumi Kubernetes Operator] requires a few different resources to install: a Service
Account, Role, RoleBinding, and Deployment, totaling about 130 lines of YAML. Let's try converting that directory
to some real code!

```sh
$ ls operator
operator.yaml  role.yaml  role_binding.yaml  service_account.yaml

> ```

Reviewer 3: Do these files exist somewhere we can link to so readers could try this out?  Or could we provide a URL to see the results in pulumi.com/kube2pulumi?

Reviewer 4: https://github.com/pulumi/pulumi-kubernetes-operator/tree/master/deploy/yaml
Agree with the above suggestion to crosslink.

Author: Linked

```typescript
const operatorName = "pulumi-kubernetes-operator";

const pulumi_kubernetes_operatorServiceAccount = new kubernetes.core.v1.ServiceAccount("pulumi_kubernetes_operatorServiceAccount", {
    apiVersion: "v1",
    kind: "ServiceAccount",
    metadata: {
        name: operatorName,
    },
});
const pulumi_kubernetes_operatorRoleBinding = new kubernetes.rbac.v1.RoleBinding("pulumi_kubernetes_operatorRoleBinding", {
    kind: "RoleBinding",
    apiVersion: "rbac.authorization.k8s.io/v1",
    metadata: {
        name: operatorName,
    },
    subjects: [{
        kind: "ServiceAccount",
        name: pulumi_kubernetes_operatorServiceAccount.metadata.name,
    }],
    roleRef: {
        kind: "Role",
        name: pulumi_kubernetes_operatorRole.metadata.name,
        apiGroup: "rbac.authorization.k8s.io",
    },
});
```

This simple refactoring is just the tip of the iceberg. At this point, you could:

- wrap these resources in a class to
abstract away the low-level details,
- add configuration values to the stack to make it easy to tweak certain values,
- or even add custom logic for Pulumi to run after a particular resource is created.

> Pulumi gives you superpowers, and

Reviewer 3: Maybe "Kubernetes superpowers"?

you decide how to use them!

## Learn More

If you'd like to learn about Pulumi and how to manage your
infrastructure and Kubernetes through code, [get started today]({{< relref "/docs/get-started" >}}). Pulumi is open
source and free to use.

For further examples on how to use Pulumi to create Kubernetes

> clusters, or deploy workloads to a cluster, check out the rest of the

Reviewer 4: Nit - Perhaps call out that we are only scratching the surface with these "super powers" and that we welcome feedback and feature requests?

---
title: "Introducing kube2pulumi: No More YAML"

> date: 2020-08-14

Reviewer: ```suggestion
date: 2020-08-13
```

> draft: true

Reviewer: ```suggestion
```
